dj/sql/parsing/ast.py

```diff
@@ -0,0 +1,517 @@
+"""
+Types to represent the DJ AST used as an intermediate representation for DJ operations
+"""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field, fields
+from enum import Enum
+from itertools import chain, zip_longest
+from typing import (
+    Any,
+    Callable,
+    Generator,
+    Generic,
+    Iterator,
+    List,
+    Optional,
+    Set,
+    Type,
+    TypeVar,
+    Union,
+)
+
+
+def flatten(maybe_iterables: Any) -> Iterator:
+    """
+    flattens `maybe_iterables` by descending into items that are Iterable
+    """
+
+    def single_item_generator(item: Any) -> Generator[Any, None, None]:
+        yield item
+
+    if not isinstance(maybe_iterables, (list, tuple, set, Iterator)):
+        return single_item_generator(maybe_iterables)
```

Comment:
`single_item_generator` could be implemented using the `iter()` builtin:

```suggestion
    if not isinstance(maybe_iterables, (list, tuple, set, Iterator)):
        return iter([maybe_iterables])
```